src/main/java/com/linecorp/armeria/common/http/Http2GoAwayListener.java

```diff
@@ -0,0 +1,84 @@
+/*
+ * Copyright 2016 LINE Corporation
+ *
+ * LINE Corporation licenses this file to you under the Apache License,
+ * version 2.0 (the "License"); you may not use this file except in compliance
+ * with the License. You may obtain a copy of the License at:
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+
+package com.linecorp.armeria.common.http;
+
+import java.nio.charset.StandardCharsets;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import io.netty.buffer.ByteBuf;
+import io.netty.buffer.ByteBufUtil;
+import io.netty.channel.Channel;
+import io.netty.handler.codec.http2.Http2Connection;
+import io.netty.handler.codec.http2.Http2ConnectionAdapter;
+import io.netty.handler.codec.http2.Http2Error;
+import io.netty.handler.codec.http2.Http2Stream;
+
+/**
+ * A {@link Http2Connection.Listener} that logs the received GOAWAY frames and makes sure disconnection.
+ */
+public class Http2GoAwayListener extends Http2ConnectionAdapter {
+
+    private static final Logger logger = LoggerFactory.getLogger(Http2GoAwayListener.class);
+
+    private final Channel ch;
+
+    public Http2GoAwayListener(Channel ch) {
+        this.ch = ch;
+    }
+
+    @Override
+    public void onGoAwaySent(int lastStreamId, long errorCode, ByteBuf debugData) {
+        onGoAway("Sent", lastStreamId, errorCode, debugData);
+    }
+
+    @Override
+    public void onGoAwayReceived(int lastStreamId, long errorCode, ByteBuf debugData) {
+        onGoAway("Received", lastStreamId, errorCode, debugData);
+        ch.close();
```

Comment:
When a GOAWAY is received, `DefaultHttp2Connection.goAwayReceived(...)` is invoked. It will notify the listeners and then close all active streams. It does not attempt to close a connection, and it seems like no known `Http2Connection.Listener.onGoAwayReceived()` implementation closes the connection either. It just waits for the client to close the connection, which doesn't sound like a good idea.